documentation/general/dotnet-run-file-multi-file.md

```diff
@@ -0,0 +1,136 @@
+# Multi-file support in `dotnet run app.cs`
+
+## Scenarios
+
+Some multi-file scenarios to consider:
+
+### Direcotry-based apps
+
+- Directory of files that are all part of the same app
+- Only a single entry-point file
+- Might be using differnt SDKs, e.g. `Microsoft.NET.Sdk` vs. `Microsoft.NET.Sdk.Web`, that have different default items
+
+Example console app:
+
+```text
+- myapp/
+  - myapp.cs
+  - helpers.cs
+```
+
+Example Web API app:
+
+```text
+- webapi/
+  - webapi.cs
+  - helpers.cs
+  - customerapi.cs
+  - productapi.cs
+  - appsettings.json
+```
+
+Example Razor app:
+
+```text
+- razorapp/
+  - razorapp.cs
+  - helpers.cs
+  - Pages/
+    - _ViewImports.razor
+    - Index.razor
+    - App.razor
+    - Layout.razor
+  - appsettings.json
+```
+
+Example Aspire app host app:
+
+```text
+- apphost/
+  - apphost.cs
+  - helpers.cs
+  - appsettings.json
+```
+
+### Directories with multiple separate entry-point files
+
+- Each entry-point file is a separate app
+- Some entry-point files require types from some non-entry-point files
+- Entry-point files don't require types from other entry-point files
+
+Example:
+
+```text
+- apps/
+  - app1.cs
+  - app2.cs
+  - webapi1.cs
+  - helpers.cs
+```
+
+### Directories with multiple related entry-point files
+
+- Each entry-point file is a separate app
+- Non-entry-point files are shared by multiple entry-point files
+- Some entry-point files require types from other entry-point files, e.g. `cat.cs` and `cat.benchmarks.cs`, `webapi.cs` and `webapi.tests.cs`
+
+Example:
+
+```text
+- apps/
+  - app1.cs
+  - app1.benchmarks.cs
+  - app2.cs
+  - app2.tests.cs
+  - helpers.cs
+```
+
+### Other considerations
+
+- Different default items for different SDKs, e.g. `Microsoft.NET.Sdk` vs. `Microsoft.NET.Sdk.Web`
+- Performance from CLI and IDE point of view
+- Workspace context in IDE for a given file (does each entry-point file map to a project? or do they all map to the same project? or do they all map to different projects?)
+
+## Approaches
+
+### Adopt default items according to the SDK being used
+
+- Defaults all files to work like a regular project
+- Can be disabled for a given entry-point file by adding a directive like `#:property EnableDefaultItems=false`
+- Can be disbled for a whole directory by adding a file `Directory.Build.props` with the following content:
+
+    ```xml
+    <Project>
+      <PropertyGroup>
+        <EnableDefaultItems>false</EnableDefaultItems>
+      </PropertyGroup>
+    </Project>
+    ```
+
+- Works for all SDKs including those that add other default items beyond **.cs* files
+- Maps naturally to converted project
+- How to know when to stop including files in the directory tree? Is this already implicitly handled given existing project-based default items behavior?
+- How to deal with multiple entry-point files?
```

Comment:
Not a blocker, but these will end up dove tailing into language design issues. Think we'd need to dive into general top level functions if end up including multiple entry point files. 

Did you consider an option where we just have the compiler exclude them? Basically if you execute `dotnet run app1.cs` that we implicitly exclude `app2.cs`?

Reply:
Yeah it seems the natural best-fit here is still to default to including default items, except for other entry-point files.

Reply:
I see like three stages here we could do incrementally:
1. Start with excluding other entry-point files.
2. Change compiler to generate Programs from top-level statements in namespaces (could be implementation-only feature flag for starters). This will simplify the generated project files - no need to have "excludes", just specify an entry-point.
3. Change language to also make local functions visible across top-level statement files.